[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor

jananisriram opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131519
* #131518

Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

jananisriram committed:
[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor

Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

davidberard98 commented on test/test_nestedtensor.py:
```diff
@@ -4312,6 +4323,132 @@ def test_softmax_reduce_batch_dim(
+        """
+        if (
+            torch._dynamo.is_compiling() and not requires_grad
+        ):  # requires_grad = False does not work with dynamo
```
note for myself: come back to check this

davidberard98 replied:
okay, this is probably fine but can you add a quick comment explaining why you need this?

what I get is
```
AssertionError: SymInts must use SymNodeVariable. If the underlying value is static, we will create a ConstantVariable and specialize.
```
from
```
   File "/data/users/dberard/pytorch/test/test_nestedtensor.py", line 4362, in torch_dynamo_resume_in_test_layer_norm_reduce_ragged_idx_1_at_4342
    *nt.shape[nt._ragged_idx + 1 :],
```

jananisriram replied:
yes definitely!

davidberard98 commented on test/test_nestedtensor.py:
```diff
@@ -4775,11 +4912,62 @@ def test_jagged_softmax_dim_with_lengths(
+                        nt_with_holes, normalized_shape=normalized_shape
+                    )
+            else:
+                out = torch.nn.functional.layer_norm(
+                    nt_with_holes, normalized_shape=normalized_shape
+                )
+
```
technically I think this will be wrong, but it's not introduced by your PR and previously was also wrong. Maybe you can file an issue or open a follow-up PR.

jananisriram replied:
opened an issue! https://github.com/pytorch/pytorch/issues/131967

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -1114,12 +1114,90 @@ def native_layer_norm_default(func, *args, **kwargs):
+            "layer_norm(): not supported for NestedTensor objects with 2 or fewer dimensions"
+        )
+
+    normalized_shape = new_kwargs[
+        "normalized_shape"
+    ]  # denote -1 as the ragged dimension, i.e. normalize on the ragged dimension
+
+    num_output_dims = inp.dim() - len(normalized_shape)
+
+    if num_output_dims == 0:  # error if trying to normalize over the batch dimension
+        raise RuntimeError(
+            "layer_norm(): not supported when normalizing over the batch dimension for NestedTensor"
+        )
+
+    if -1 in normalized_shape and inp._lengths is not None:
+        raise RuntimeError(
+            "layer_norm(): not supported where lengths is not None if operating on the ragged dimension for NestedTensor"
+        )
+
+    if (
+        -1 in normalized_shape
+    ):  # special handling for normalizing over the ragged dimension
```
Could you change the handling here?

-1 makes sense, but we could also get a nested int here, e.g. j0 or something.

I feel like it might be better to say "if len(normalized_shape) == len(self._size) - 1` (not sure if that's the right way to get the size).

Also, I think technically there's another case for `len(normalized_shape) == len(self._size)`, which isn't handled here. I think it's fine to not handle it but maybe add an error.

jananisriram replied:
> -1 makes sense, but we could also get a nested int here, e.g. j0 or something.

i was mixing up NSTs (which don't support shape on the ragged dimension) with NJTs (which do support it); thanks for pointing this out!

> I feel like it might be better to say "if len(normalized_shape) == len(self._size) - 1` (not sure if that's the right way to get the size).

i think this would need to be changed in the future when we handle `jt_all` (`ragged_idx > 1`). for example, if (in the future) we try to normalize a nt of shape `(4, 5, j0, 6, 7)` on `(5, j0, 6, 7)`, checking the lengths would become inaccurate!

> there's another case for len(normalized_shape) == len(self._size)

would this be the same as checking `inp.dim() - len(normalized_shape) == 0` (aka normalizing on batch dimension)? if so, i think this is handled and throws an error!

also, changed `num_output_dims` to a variable name that's more descriptive :)

davidberard98 reviewed (approved):
LGTM, can you take a look at the index handling though?

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

Differential Revision: [D59969580](https://our.internmc.facebook.com/intern/diff/D59969580/)

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

[ghstack-poisoned]

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

[ghstack-poisoned]

jananisriram commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "broke CI: test_nestedtensor.py::TestNestedTensorSubclassCPU::test_layer_norm_with_lengths_requires_grad_False_components_require_grad_False_cpu_float32 [GH job link](https://github.com/pytorch/pytorch/actions/runs/10121747545/job/27996722731) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/f862f457304f1952e75336f9f74e4ea3d2a5eb72)"

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

[ghstack-poisoned]

davidberard98 commented on test/test_nestedtensor.py:
```diff
@@ -4928,6 +4928,12 @@ def test_layer_norm_with_lengths(
+        # requires_grad = False does not currently work with dynamo tests and throws this error:
+        # AssertionError: SymInts must use SymNodeVariable. If the underlying value is static,
+        # we will create a ConstantVariable and specialize.
+        if torch._dynamo.is_compiling():
```
should this also check for `not requires_grad`? or does the comment need to be updated?

If we're always skipping dynamo then I think it would be better to just xfail or skip if dynamo (with an annotation or with a file in test/dynamo_expected_failures or test/dynamo_skips)

jananisriram replied:
test might be flaky, but it recently started to throw errors for both `not requires_grad` and `requires_grad` (though it wasn't throwing errors for `requires_grad` before. i'll update the comment and add the appropriate decorator!

jananisriram committed:
Update on "[NestedTensor] Integrate the layer normalization operator along the jagged dimension into NestedTensor"


Modify the existing `layer normalization` operator in PyTorch, invoked by `torch.layer_norm`, to allow for reductions along the jagged dimension of a nested tensor. The function originally had a basic implementation for reducing along 1 non-ragged dimension. This diff, which uses the `aten` padding operator, enables PyTorch users to invoke `torch.nn.functional.layer_norm` on a nested tensor when reducing along the ragged dimension, e.g. `*` in a `(B, *, M)` or `(B, *, M, N)` nested tensor.

Write unit tests based on the `softmax` jagged operator to verify the accuracy of the ragged reduction implementation for `torch.nn.functional.layer_norm`. Add unit tests to verify error handling for unsupported features.

Note that this implementation is limited to nested tensors with `ragged_idx == 1`, i.e. the ragged dimension is not transposed. The layer normalization operator also requires an operation on a 2-dimensional layer; for nested tensors with 4 or more dimensions, I flatten the extra dimensions, then unflatten them after performing layer normalization.

[ghstack-poisoned]

jananisriram commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c weird -m "Sorry, reverting this since this is based on an internal diff that has diverged from actual internal commit.  Weird conflicts arise when that happens.  Let's revert both this PR and the internal diff, and then reland them as a proper new codev diff"

ZainRizvi reviewed (changes requested):
Note: Please do not merge this PR.  Create a new codev diff and only use the codev tools to push the changes to github.  Alternatively, you can create the new PR directly on github and the pytorch diff train will take care of get it merged internally
